versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/AggBsonUpdateVisitor.java

```diff
@@ -0,0 +1,107 @@
+ */
+class AggBsonUpdateVisitor implements UpdateExpressionVisitor<List<Bson>> {
+  private static class StageOp {
+    final String path;
+    final String index;
+    final String operation;
+
+    StageOp(String path, String index, String operation) {
+      this.path = path;
+      this.index = index;
+      this.operation = operation;
+    }
+  }
+
+  private static class ClauseVisitor implements UpdateClauseVisitor<StageOp> {
+    @Override
+    public StageOp visit(RemoveClause clause) {
+      final String path = clause.getPath().getRoot().accept(BsonPathVisitor.INSTANCE_NO_QUOTE, true);
+      if (isArrayPath(clause.getPath())) {
+        // Mongo has no way of deleting an element from an array, it by default leaves a NULL element. To remove the
+        // element in one operation, splice together the array around the index to remove.
+        final int lastIndex = path.lastIndexOf(".");
+        final String arrayPath = path.substring(0, lastIndex);
+        final String arrayIndex = path.substring(lastIndex + 1);
+        return new StageOp(arrayPath, arrayIndex, String.format("{$addFields: {\"%1$s\": {$concatArrays: [{$slice:[ \"$%1$s\", %2$s]}, "
+            + "{$slice:[ \"$%1$s\", {$add:[1, %2$s]}, {$size:\"$%1$s\"}]}]}}}",
+            arrayPath, arrayIndex));
+      }
+
+      return new StageOp(path, "", String.format("{$project: {\"%s\": 0}}", path));
+    }
+
+    @Override
+    public StageOp visit(SetClause clause) {
+      throw new UnsupportedOperationException(String.format("Unsupported SetClause type: %s", clause.getValue().getType().name()));
+    }
+  }
+
+  private static final ClauseVisitor CLAUSE_VISITOR = new ClauseVisitor();
+
+  @Override
+  public List<Bson> visit(UpdateExpression expression) {
+    // Sort the operations so the array removals occur in reverse order to avoid indexing problems. Since we use a pipeline,
+    // if they occur in increasing order then consider the case of: [a, b, c] with a removal of 0, 1. You would expect
+    // a return array of [c], however what occurs is actually:
+    //  - remove 0: [a, b, c] -> [b, c]
+    //  - remove 1: [b, c] -> [b]
+    // if instead they occur in reverse order:
+    //  - remove 1: [a, b, c] -> [a, c]
+    //  - remove 0: [a, c] -> [c]
+    // TODO: opportunity to optimize the removals by collapsing sequential removed indexes into ranges.
+    return expression.getClauses().stream()
+      .map(c -> c.accept(CLAUSE_VISITOR))
+      .sorted((o1, o2) -> {
+        final int compare = o2.path.compareTo(o1.path);
+        return (compare == 0) ? o2.index.compareTo(o1.index) : compare;
+      })
+      .map(o -> BsonDocument.parse(o.operation))
+      .collect(Collectors.toList());
+  }
+
+  /**
+   * Determine whether or not the given ExpressionPath refers to an array element.
+   * @param path the path to check.
+   * @return true if the path is for an array element, false otherwise.
+   */
+  static boolean isArrayPath(ExpressionPath path) {
```

Comment:
Since we know we're testing a path, does it need to be in the function name? Better if the function name only refers to what is being tested. Unless it's an array path, of course.
Can it be "isArray" only? or "isBsonArray"?